java/core/src/main/java/com/whylogs/core/DatasetProfile.java

```diff
@@ -0,0 +1,129 @@
+package com.whylogs.core;
+
+import com.whylogs.core.metrics.Metric;
+import com.whylogs.core.schemas.ColumnSchema;
+import com.whylogs.core.schemas.DatasetSchema;
+import com.whylogs.core.views.ColumnProfileView;
+import com.whylogs.core.views.DatasetProfileView;
+import java.time.ZonedDateTime;
+import java.util.*;
+import lombok.Getter;
+import lombok.ToString;
+
+// TODO: extend WRITABLE interface
+@Getter
+@ToString
+public class DatasetProfile {
+  // TODO: Time zone is all mixed up. Fix
+  public static int _LARGE_CACHE_SIZE_LIMIT = 1024 * 100;
+
+  private DatasetSchema schema;
+  // QUESTION: Do we need zones here? Do we just use UTC?
+  private Date datasetTimestamp;
+  private Date creationTimestamp;
+  private HashMap<String, ColumnProfile<?>> columns;
+  private boolean isActive = false;
+  private int trackCount = 0;
+
+  public DatasetProfile(
+      Optional<DatasetSchema> datasetSchema,
+      Optional<Date> datasetTimestamp,
+      Optional<Date> creationTimestamp) {
+    this.schema = datasetSchema.orElse(new DatasetSchema());
+    this.datasetTimestamp = datasetTimestamp.orElse(new Date());
+    this.creationTimestamp = creationTimestamp.orElse(new Date());
+
+    this.columns = new HashMap<>();
+    this.initializeNewColumns(schema.getColNames());
+  }
+
+  public void addMetric(String colName, Metric metric) {
+    if (!this.columns.containsKey(colName)) {
+      throw new InputMismatchException("Column name not found in schema");
+    }
+    this.columns.get(colName).addMetric(metric);
+  }
+
+  public void track(HashMap<String, Object> row) {
+    try {
+      this.isActive = true;
+      this.trackCount += 1;
+      this.doTrack(row);
+    } finally {
+      this.isActive = false;
+    }
+  }
+
+  private void doTrack(HashMap<String, Object> row) {
+    boolean dirty = this.schema.resolve(row);
+    if (dirty) {
+      Set<String> schemaColumnNames = this.schema.getColNames();
+      Set<String> newColumnNames = new HashSet<>();
+      for (String colName : schemaColumnNames) {
+        if (!this.columns.containsKey(colName)) {
+          newColumnNames.add(colName);
+        }
+      }
+      this.initializeNewColumns(newColumnNames);
+    }
+
+    for (String col : row.keySet()) {
+      ArrayList<Object> values = new ArrayList<>();
```

Comment:
Maybe pull `values` up to a top-level member to avoid the alloc/construction/destruction in the loop?